File: DSA/src/Questions/BackTracking/JudgePoint24.java
package Questions.BackTracking;

/*
https://leetcode.com/problems/24-game/description/
 You are given an integer array cards of length 4. You have four cards, each containing a number in the range [1, 9]. You should arrange the numbers on these cards in a mathematical expression using the operators ['+', '-', '*', '/'] and the parentheses '(' and ')' to get the value 24.

You are restricted with the following rules:

The division operator '/' represents real division, not integer division.
For example, 4 / (1 - 2 / 3) = 4 / (1 / 3) = 12.
Every operation done is between two numbers. In particular, we cannot use '-' as a unary operator.
For example, if cards = [1, 1, 1, 1], the expression "-1 - 1 - 1 - 1" is not allowed.
You cannot concatenate numbers together
For example, if cards = [1, 2, 1, 2], the expression "12 + 12" is not valid.
Return true if you can get such expression that evaluates to 24, and false otherwise.

 

Example 1:

Input: cards = [4,1,8,7]
Output: true
Explanation: (8-4) * (7-1) = 24
Example 2:

Input: cards = [1,2,1,2]
Output: false
 

Constraints:

cards.length == 4
1 <= cards[i] <= 9
Time Complexity: O(6^N)
Space Complexity: O(1)
 */
public class JudgePoint24 {
    private static final double EPS = 1e-6;

    private boolean backtrack(double[] A, int n) {
        if (n == 1)
            return Math.abs(A[0] - 24) < EPS;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                double a = A[i], b = A[j];
                A[j] = A[n - 1];

                A[i] = a + b;
                if (backtrack(A, n - 1))
                    return true;

                A[i] = a - b;
                if (backtrack(A, n - 1))
                    return true;

                A[i] = b - a;
                if (backtrack(A, n - 1))
                    return true;
                A[i] = a * b;
                if (backtrack(A, n - 1))
                    return true;

                A[i] = a / b;
                if (backtrack(A, n - 1))
                    return true;

                A[i] = b / a;
                if (backtrack(A, n - 1))
                    return true;

                A[i] = a;
                A[j] = b;
            }
        }
        return false;
    }

    public boolean judgePoint24(int[] nums) {
        double[] A = new double[nums.length];
        for (int i = 0; i < nums.length; i++)
            A[i] = nums[i];
        return backtrack(A, A.length);
    }
    public static void main(String[] args) {
        JudgePoint24 judgePoint24 = new JudgePoint24();
        System.out.println(judgePoint24.judgePoint24(new int[]{4,1,8,7}));
        System.out.println(judgePoint24.judgePoint24(new int[]{1,2,1,2}));
    }
}

File: DSA/Python/Questions/BackTracking/JudgePoint24.py
"""
# Question: 24 Game
# Link: https://leetcode.com/problems/24-game/

# Find if it's possible to make 24 using 4 numbers with basic operations

# Time Complexity: O(6^N)
# Space Complexity: O(1)

# Algorithm:
# 1. Try all possible pairs of numbers
# 2. Apply all possible operations (+,-,*,/)
# 3. Recursively solve with remaining numbers
# 4. Check if result equals 24

# Key Components:
# - backtrack(): Recursive helper for operation combinations
# - judge_point_24(): Main implementation
# - Floating point comparison handling
"""


class JudgePoint24:
    def __init__(self):
        self.EPS = 1e-6

    def backtrack(self, A: list, n: int) -> bool:
        if n == 1:
            return abs(A[0] - 24) < self.EPS

        for i in range(n):
            for j in range(i + 1, n):
                a, b = A[i], A[j]
                A[j] = A[n - 1]

                # Try all operations
                A[i] = a + b
                if self.backtrack(A, n - 1):
                    return True

                A[i] = a - b
                if self.backtrack(A, n - 1):
                    return True

                A[i] = b - a
                if self.backtrack(A, n - 1):
                    return True

                A[i] = a * b
                if self.backtrack(A, n - 1):
                    return True

                if b != 0:
                    A[i] = a / b
                    if self.backtrack(A, n - 1):
                        return True

                if a != 0:
                    A[i] = b / a
                    if self.backtrack(A, n - 1):
                        return True

                A[i], A[j] = a, b

        return False

    def judge_point_24(self, nums: list) -> bool:
        A = [float(num) for num in nums]
        return self.backtrack(A, len(A))


def main():
    solution = JudgePoint24()
    print(solution.judge_point_24([4, 1, 8, 7]))  # True
    print(solution.judge_point_24([1, 2, 1, 2]))  # False


if __name__ == "__main__":
    main()
